### ISSUE TITLE
crashlytics didCrashDuringPreviousExecution always returns false for macOS

### ISSUE THREAD

**Issue by saurabhprasad** [firebase/firebase-ios-sdk#10458]
### Description

We recently onboarded to firebase version 8.10 for crashlytics support in our macOS app. It is basically an iPad app running on M1 silicon. 
What we observed is even though the crashes are getting reported and execution works fine, the crashlytics api `didCrashDuringPreviousExecution` always returns false.
We verified that when crash occurs the `previously-crashed` does gets generated. The same issue is not observed when the app runs on iPad.

### Reproducing the issue

Yes

### Firebase SDK Version

8.10

### Xcode Version

13.4.1

### Installation Method

CocoaPods

### Firebase Product(s)

Crashlytics

### Targeted Platforms

macOS

### If using Swift Package Manager, the project's Package.resolved

<details>
<summary>Expand <code>Package.resolved</code> snippet</summary>
<br>

```json

Replace this line with the contents of your Package.resolved.

```

</details>

### If using CocoaPods, the project's Podfile.lock

<details>
<summary>Expand <code>Podfile.lock</code> snippet</summary>
<br>

```yml

Replace this line with the contents of your Podfile.lock!

```

</details>

**Comment by rizafran** [comment]
Thanks for reporting, @saurabhprasad. Does updating the SDK to a later version makes difference?

**Comment by saurabhprasad** [comment]
hi @rizafran thanks for your response, I can try updating it but if there is no changes done in that direction then i guess it should be reproducible on latest version as well.

**Comment by Deco354** [comment]
We've got this issue as well. We were using it to send analytics to one of our internal analytic tools and we've 0 crashes across the board now.

**Comment by andrewdmontgomery** [comment]
We are seeing the same thing, and we're seeing it across nearly all of our iOS devices.  In our case, we were doing something like this:

```swift
FirebaseApp.configure()
if Crashlytics.crashlytics().crashlytics.didCrashDuringPreviousExecution() {
    // do stuff
}
```

If we set a breakpoint on line 2, we discovered that the Crashlytics singleton (`Crashlytics.crashlytics()`) was still uninitialized.  By the `FIRCrashlytics` class uses the macro [NS_ASSUME_NONULL_BEGIN](https://github.com/firebase/firebase-ios-sdk/blob/d6e18b978aab37c810a6e9ce40a756ae02e58562/Crashlytics/Crashlytics/Public/FirebaseCrashlytics/FIRCrashlytics.h#L26) (and end).  I'm not very familiar with that, but it seems that Swift interoperability is still allowing me to send a message to this `null` object, which is returning `false`.

If I make changes that allow the singleton to finish initializing, the call to `didCrashDuringPreviousExecution()` returns correctly.

Note that we had `FirebaseCrashlyticsCollectionEnabled` set to `true` in our `Info.plist`, which means that our call to `.didCrashDuringPerviousExecution()` was actually experiencing a double-race condition:
- If we call it too soon, the singleton was still uninitialized and it was always returning `false`
- If we call it too late, Crashlytics will already have sent the logs and cleaned up the `previous-crash` evidence (so it will correctly return `false`)

I tested this on Firebase 9.5.0 and 9.6.0.

It is a bit of a mystery as to why this started happening to us though.  We have made no changes to the code that is directly responsible for this, and we haven't updated Firebase.  But given the double-race condition, my best guess is that something unrelated changed, and inadvertently changed the timing of these two events.

**Comment by andrewdmontgomery** [comment]
We identified what was happening to us.  We have a class that is keeping a reference to the `Crashlytics.crashlytics()` singleton in a private property.  That reference was assigned during the initialization of the class.  For example:

```swift
class MyClass {
    private let crashlytics = Crashlytics.crashlytics()

    func doStuff() {
        let didCrash = craslytics.didCrashDuringPreviousExecution()  // returns false if `crashlytics` is "uninitialized
        if didCrash {
            crashlytics.sendUnsentReports()
        }
    }
}
```

The problem was that we changed how (and when) we were calling `didCrashDuringPreviousExecution()` by calling it in a class that gets initialized earlier in the app launch:
1. Initialize `MyClass` during `willFinishLaunchingWithOptions` in `AppDelegate`
2. Call `FirebaseApp.configure()` during `didFinishLaunchingWithOptions` in `AppDelegate`

As a result, that `crashlytics` property has a reference to the "uninitialized" Crashlytics singleton.

Our quick fix was to move `FirebaseApp.configure()` into `willFinishLaunchingWithOptions`, before we initialize the class that depends on the singleton.

But we are going to refactor this substantially.  Depending on how your situation, it may be sufficient to simply change that stored property to a computed property:

```swift
class MyClass {
    private let crashlytics: Crashlytics { Crashlytics.crashlytics() }

    func doStuff() {
        let didCrash = craslytics.didCrashDuringPreviousExecution()
        if craslytics.didCrashDuringPreviousExecution() {
            crashlytics.sendUnsentReports()
        }
    }
}
```

But you'd still need to guarantee that you called `FirebaseApp.configure()` before referencing that property.
